pkg/client/bucket/bucket.go

```diff
@@ -0,0 +1,210 @@
+		return r, nil
+	}
+
+	log.WithCtxFields(ctx).Debug("failed to create new object with ID %q; trying to update existing: %w", id, err)
+
+	b, err := c.get(ctx, id)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to get object with ID %q: %w", id, err)
+	}
+
+	r, err = c.update(ctx, b, data)
+	log.WithCtxFields(ctx).Debug("updated object with ID %q", id)
+	return r, err
+}
+
+func (c Client) create(ctx context.Context, id string, data []byte) (Response, error) {
+	u, err := url.JoinPath(c.url, endpoint)
+	if err != nil {
+		return Response{}, fmt.Errorf("faild to create sound url: %w", err)
+	}
+
+	err = setBucketName(id, &data)
+	if err != nil {
+		return Response{}, err
+	}
+
+	r, err := c.client.Post(ctx, u, data)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to create object with ID %q: %w", id, err)
+	}
+	if !r.IsSuccess() {
+		return Response{}, rest.NewRespErr(fmt.Sprintf("failed to update object with ID %q (HTTP %d): %s", id, r.StatusCode, string(r.Body)), r).WithRequestInfo(http.MethodPut, u)
+	}
+
+	b, err := unmarshalJSON(r.Body)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to read response: %w", err)
+	}
+
+	return b, nil
+}
+
+func (c Client) update(ctx context.Context, b Response, data []byte) (Response, error) {
+
+	u, err := url.Parse(c.url)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to parse url: %w", err)
+	}
+
+	u.Path, err = url.JoinPath(u.Path, endpoint, b.BucketName)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to join url: %w", err)
+	}
+
+	q := u.Query()
+	q.Add("optimistic-locking-version", strconv.Itoa(b.Version))
+	u.RawQuery = q.Encode()
+
+	var m map[string]any
+	err = json.Unmarshal(data, &m)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to unmarshal template: %w", err)
+	}
+	m["bucketName"] = b.BucketName
+	m["version"] = b.Version
+	m["status"] = b.Status
+
+	data, err = json.Marshal(m)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to marshal data: %w", err)
+	}
+
+	r, err := c.client.Put(ctx, u.String(), data)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to update object with ID %q: %w", b.BucketName, err)
+	}
+	if !r.IsSuccess() {
+		return Response{}, rest.NewRespErr(fmt.Sprintf("failed to update object with ID %q (HTTP %d): %s", b.BucketName, r.StatusCode, string(r.Body)), r).WithRequestInfo(http.MethodPut, u.String())
+	}
+
+	return b, nil
+}
+
+func setBucketName(bucketName string, data *[]byte) error {
+	var m map[string]interface{}
+	err := json.Unmarshal(*data, &m)
+	if err != nil {
+		return err
+	}
+	m["bucketName"] = bucketName
+	*data, err = json.Marshal(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c Client) get(ctx context.Context, id string) (Response, error) {
+	u, err := url.JoinPath(c.url, endpoint, id)
+	if err != nil {
+		return Response{}, fmt.Errorf("faild to create sound url: %w", err)
+	}
+
+	retry := rest.RetrySetting{
```

Comment:
What lead to adding retries to the GET but none to POST and PUT calls for this API/client?

Reply:
As mentioned before, we had issues in our first approach in fetching buckets after the creation. We changed the code in a way that it's currently not necessary to have the retry, but keeping it would also not hurt too much.